src/snap/proxy/stream_collector_proxy.cc

```diff
@@ -0,0 +1,304 @@
+        std::string task_id = "not-set";
+
+        std::vector<Metric> send_mets, recv_mets;
+        std::string err_msg;
+
+        auto sendch = std::async(std::launch::async, &StreamCollectorImpl::metricSend,
+                                this, task_id, context, stream);
+        auto recvch = std::async(std::launch::async, &StreamCollectorImpl::streamRecv,
+                                this, task_id, context, stream);
+        auto errch = std::async(std::launch::async, &StreamCollectorImpl::errorSend,
+                                this, context, stream);
+
+        auto do_puts = std::async(std::launch::async, &StreamCollectorImpl::PutSendMetsAndErrMsg,
+                                this, context);
+
+        _stream_collector->stream_metrics(send_mets, recv_mets, err_msg);
+
+        return Status::OK;
+    } catch(PluginException &e) {
+        return Status(StatusCode::UNKNOWN, e.what());
+    }
+}
+
+bool StreamCollectorImpl::PutSendMetsAndErrMsg(ServerContext* context) {
+    while(!context->IsCancelled()) {
+        if (_stream_collector->put_mets()) {
+            _sendChan.put(_stream_collector->put_metrics_out());
+            _stream_collector->set_put_mets(false);
+        }
+        if (_stream_collector->put_err()) {
+            _errChan.put(_stream_collector->put_err_msg());
+            _stream_collector->set_put_err(false);
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(500));
+    }
+}
+
+bool StreamCollectorImpl::errorSend(ServerContext* context,
+                                    ServerReaderWriter<CollectReply, CollectArg>* stream) {
+    //std::cout << "Starting routine for sending errors" << std::endl;      
+    try {
+        CollectReply errReply;
+        ErrReply er;
+        std::string err;
+        while (context->IsCancelled()) {
+            if (_errChan.get(err)) {
+                er.set_error(err);
+                errReply.set_allocated_error(&er);
+                stream->Write(errReply);
+                err.clear();
+            }
+        }
+        return true;
+    } catch (PluginException &e) {
+        std::cout << "Error" << std::endl;
+        return false;
+    }
+}
+
+bool StreamCollectorImpl::metricSend(const std::string &taskID,
+                                    ServerContext* context,
+                                    ServerReaderWriter<CollectReply, CollectArg>* stream) {
+    MetricsReply* resp;
+    std::vector<Metric> send_mets;
+    //std::cout << "Starting routine for sending metrics" << std::endl;    
+    try {
+        std::chrono::system_clock::time_point start = std::chrono::system_clock::now();
+        while (!context->IsCancelled()) {
+            if (_sendChan.get(send_mets)) {
+                if (!send_mets.empty()) {
+                    for (Metric met : send_mets) {
+                        *resp->add_metrics() = *met.get_rpc_metric_ptr();
+
+                        if (resp->metrics_size() == _max_metrics_buffer) {
```

Comment:
I think this condition should include (_max_metrics_buffer == 0). _max_metrics_buffer set to 0 means that metric should be streamed immediately. In current implementation it won't behave proper way